File: proxy-gateway-core/src/cheap_random.rs
//! Fast, good-enough random number generation for non-cryptographic use.

use std::cell::Cell;

/// Fast, good-enough random using a thread-local xorshift64.
///
/// Seeded from the current time and thread ID so different threads and
/// different runs produce different sequences. Not suitable for cryptographic
/// use — only for tie-breaking and session ID generation.
pub fn cheap_random() -> u64 {
    thread_local! {
        static STATE: Cell<u64> = Cell::new({
            let t = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_nanos() as u64;
            let tid = std::thread::current().id();
            let tid_bits = format!("{:?}", tid);
            let tid_hash = tid_bits
                .bytes()
                .fold(0u64, |acc, b| acc.wrapping_mul(31).wrapping_add(b as u64));
            t ^ tid_hash ^ 0x517cc1b727220a95
        });
    }
    STATE.with(|s| {
        let mut x = s.get();
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        s.set(x);
        x
    })
}

File: proxy-gateway/core/cheap_random.go
package core

import (
	"sync"
	"time"
)

// CheapRandom returns a fast non-cryptographic random uint64.
// Uses a pool of xorshift64 states, one per goroutine slot.
// Seeded from time so different runs produce different sequences.
// Not suitable for cryptographic use — only for tie-breaking.

type xorshift64State struct {
	x uint64
}

func (s *xorshift64State) next() uint64 {
	s.x ^= s.x << 13
	s.x ^= s.x >> 7
	s.x ^= s.x << 17
	return s.x
}

var (
	rngPool = sync.Pool{
		New: func() interface{} {
			seed := uint64(time.Now().UnixNano()) ^ 0x517cc1b727220a95
			if seed == 0 {
				seed = 0x517cc1b727220a95
			}
			return &xorshift64State{x: seed}
		},
	}
)

// CheapRandom returns a random uint64. Fast and non-cryptographic.
func CheapRandom() uint64 {
	s := rngPool.Get().(*xorshift64State)
	v := s.next()
	rngPool.Put(s)
	return v
}
